ADR/0004-component-image-location.md

```diff
@@ -0,0 +1,187 @@
+| ----------- | ----------- |
+| Quay.io organization      | quay.io/redhat-user-workloads       |
+| OAuth Application  name | Created, name redacted        |
+| Account used to generate token | `shbose` |
+| Scope | Administer organizations, Adminster repositories, Create Repositories |
+
+<img width="1313" alt="image" src="https://user-images.githubusercontent.com/545280/212758440-5807cd4e-11b1-43bc-aa03-385b9284cb9e.png">
+
+
+#### Organization and Image Repository creation
+
+When a user creates a Component, a StoneSoup service would need to generate the image repository for consumption by the 
+build, test and deployment services.
+
+* For each user, create a new Quay.io org “quay.io/unique-org-david”
+* For each `Component` 'foo', create a new image repo “quay.io/unique-org-david/appname/componentname”
+* Configure the robot account “redhat-<resource-uuid>” in “quay.io/unique-org-david” to be able to push to “quay.io/unique-org-david/appname-foo” 
+* Configure a `Secret` in the user's namespace with the robot account token.
+* Annotate the `Component` with the name of the image repository and the name of the `Secret` containing the robot account token.
+
+The following deviations from this design would be implemented:
+
+Until the capability to progammatically create organizations in Quay.io is activated:
+* Images repositories would be created in quay.io/redhat-user-workloads
+* Isolation would continue to be maintained the same way - every image repository would have a scoped robot account that would be rotatable using Quay.io API for the same https://docs.quay.io/api/swagger/#!/robot/regenerateUserRobotToken.
+
+Until the capability to determine the associated user/tenant/Space a Component is implemented,
+* The quay.io repository created with use the format   quay.io/org-name/namespace-name/application-name/component-name. Example, https://quay.io/repository/redhat-user-workloads/shbose/applicationname/automation-repo
+
+#### Lifecycle of the Quay.io resources
+
+* Token generation: 
+    * Robot account token: At the moment, the controller responsible for generating the Quay.io resources would be responsible for rotating the tokens https://docs.quay.io/api/swagger/#!/robot/regenerateUserRobotToken
+    * Quay.io API token: No programmatic way to regenerate this token is known at this point of time. This would be a manual activity to begin with. 
+
+* Upon deletion of an `Application`/`Component` from StoneSoup, 
+    * The controller/finalizer would delete the the relevant Quay.io resources namely, the image repository and the robot account.
+    * The controller/finalizer would delete the linked `Secret` from the user's namespace. Most likely, this should be a mere `ownerReference`-based garbage collection.
+
+* Upon removal of a user from Stonesoup,
+    * The empty Quay.io organization associated with the user or the user's Space may not be deleted instantly, but would be scheduled for a delayed cleanup.
```

Comment:
>  Preventing others from being able to squat on it after cleanup?

Yes, this is what I had in mind.

Reply:
Gotcha, I wasn't sure if you had any other reason in mind. It doesn't seem like a necessary feature to protect against the quay repo itself (and potentially introduce unnecessary complexity). I imagine that it would be good to have a "delete process" where all the content which will be deleted is itemized so that the user can make an informed decision.

I also assume that any creation of organizations will be able to handle "fuzzing" to resolve name conflicts so there wouldn't be an issue if someone did manage to squat on an organization..

Is there any planned differentiation between empty and nonempty quay organizations?